### PR TITLE
add ensure_binary/str/text helper functions

### PR DESCRIPTION
**Pull request by jz1371**
* Summary
add ensure_binary, ensure_str, ensure_text helper functions to solve the encoding compatibility
between py2 and py3 more easily.
-- ensure_binary, input --> six.binary_type
-- ensure_str, input -> six.string_type
-- ensure_text, input -> six.text_type
This is mentioned in Lisa Guo's keynote in Pycon 2017 (from 31:21 in https://www.youtube.com/watch?v=66XoCk79kjM), which helped Instagram to upgrade to Python 3 and I hope this can help other developers, too

* Test Plan
add new unit tests to cover added code
```
 195 passed, 2 skipped in 0.48 seconds 
```

### CODE REVIEW

File: six.py [hunk 1]
```diff
@@ -843,6 +843,63 @@ def wrapper(cls):
...
+    if not isinstance(s, (text_type, binary_type)):
+        raise TypeError("not expecting type '%s'" % type(s))
+    if isinstance(s, text_type):
+        return s.encode(encoding, errors)
+    else:
+        return s
+
+
+def ensure_str(s, encoding='utf-8', errors='strict'):
+    """ A helper function to ensure output is six.str_type.
```
Comment (benjaminp) [inline six.py]: `six.str_type` doesn't exist.

File: six.py [hunk 2]
```diff
@@ -843,6 +843,63 @@ def wrapper(cls):
...
+
+    For Python 3:
+      - `str` -> encoded to `bytes`
+      - `bytes` -> `bytes`
+    """
+    if not isinstance(s, (text_type, binary_type)):
+        raise TypeError("not expecting type '%s'" % type(s))
+    if isinstance(s, text_type):
+        return s.encode(encoding, errors)
+    else:
```
Comment (benjaminp) [inline six.py]: This can be `elif isinstance(s, binary_type):` and then the first `if` statement in this function may be removed.

File: six.py [hunk 3]
```diff
@@ -843,6 +843,63 @@ def wrapper(cls):
...
+
+    For Python 3:
+      - `str` -> `str`
+      - `bytes` -> decoded to `str`
+    """
+    if not isinstance(s, (text_type, binary_type)):
+        raise TypeError("not expecting type '%s'" % type(s))
+    if isinstance(s, binary_type):
+        return s.decode(encoding, errors)
+    else:
```
Comment (benjaminp) [inline six.py]: The `isinstance` checks can be rewritten in a way similar to what I suggested in `ensure_binary`.

**Comment by jz1371** [comment]
Thanks for the review, @benjaminp 
- for `doc`, I plan to add that after function `six.iterbytes(buf)`  in section https://pythonhosted.org/six/#binary-and-text-data, i.e. 

screen shot 2017-08-06 at 12 31 12 pm
  please let me know if you have a better suggestion

- sure, will add tests to cover arguments `errors`, `encoding`
- and test for verifying the return type

File: six.py (hunk 2 above)
Comment (jz1371) [inline six.py]: @benjaminp do you mean something like?
```
def ensure_binary(s, encoding='utf-8', errors='strict'):
    if isinstance(s, text_type):
        return s.encode(encoding, errors)
    elif isinstance(s, binary_type):
        return s
    else:
        raise TypeError("not expecting type '%s'" % type(s))
```
I'd like to keep the `raise TypeError` part to fail the error loudly; be consistent with the `raise` behavior from `ensure_string`.

File: six.py (hunk 1 above)
Comment (jz1371) [inline six.py]: good catch, will update the doc to six.string_types and rename function to `ensure_string`

File: six.py (hunk 3 above)
Comment (jz1371) [inline six.py]: sure, will update to the following if we agree to keep `raise exception` part
```
    if isinstance(s, binary_type):
        return s.decode(encoding, errors)
    elif isinstance(s, text_type):
        return s
    else:
        raise TypeError("not expecting type '%s'" % type(s))
```

File: documentation/index.rst [hunk 4]
```diff
@@ -433,6 +433,24 @@ string data in all Python versions.
    a bytes object iterator in Python 3.
 
 
+.. function:: ensure_binary(s, encoding='utf-8', errors='strict')
+
+  A helper function to ensure output is :data:`binary_type`. ``encoding``, ``errors``
+  are the same as https://docs.python.org/3/library/stdtypes.html#str.encode
```
Comment (benjaminp) [inline documentation/index.rst]: Write
```
:meth:`py3:str.encode`
```
here and below

File: test_six.py [hunk 5]
```diff
@@ -903,3 +903,56 @@ def __bytes__(self):
         assert str(my_test) == six.u("hello")
 
     assert getattr(six.moves.builtins, 'bytes', str)(my_test) == six.b("hello")
+
+
+class EnsureTests(unittest.TestCase):
```
Comment (benjaminp) [inline test_six.py]: Don't need to inherit from `TestCase` here.

**Comment by mariocj89** [comment]
Hi! @benjaminp, @jz1371 is it moving forward? @jz1371 if you have no time I am happy to finish it.

You can see there is a need for these functions in the community. 

They even made it to a keynote at PyCon 🤣 https://youtu.be/66XoCk79kjM?t=31m15s

**Comment by jz1371** [comment]
@benjaminp, any other changes you'd like to have in this PR?

**Comment by cclauss** [comment]
THIS IS SO AWESOME!!  I was watching the Pycon 2017 Instagram video and thinking "This needs to be in six!" and it is already.  Any sense on when this will be in a release?

**Comment by mariocj89** [comment]
Indeed. I am looking forward to remove this from m codebase :)
@benjaminp do you plan to cut a release or is there any way we can help you do it?

**Comment by anthrotype** [comment]
It would be great if we could have this released to pypi some time soon, thanks!

**Comment by ofek** [comment]
@benjaminp Hello there! Any idea when this may be released?

**Comment by JinmingZhao** [comment]
six==1.14 will work with the error 'module 'six' has no attribute 'ensure_str''

**Comment by cclauss** [comment]
six==1.14.0 (with the __.0__) as in https://pypi.org/project/six/
* https://github.com/benjaminp/six/releases

**Pull request merged**